Revit_Core_Engine/Query/LocationCurve.cs

```diff
@@ -140,6 +145,182 @@ public static BH.oM.Geometry.Line LocationCurveColumn(this FamilyInstance family
         }
 
         /***************************************************/
+        
+        public static List<BH.oM.Geometry.Line> LocationCurveMEP (this MEPCurve mepCurve, RevitSettings settings = null)
+        {
+            //List<BH.oM.Geometry.Line> result = new List<BH.oM.Geometry.Line>();
+            settings = settings.DefaultIfNull();
+            
+            // Determine start and end points, which if the MEP linear object is connected to
+            // a fitting then uses the fitting origin as either start/end
+            List<BH.oM.Geometry.Point> allPoints = new List<BH.oM.Geometry.Point>();
+            
+            var tolerance = BH.oM.Adapters.Revit.Tolerance.Vertex.FromSI(UnitType.UT_Length);
+            
+            BH.oM.Geometry.Line extraLine1 = null;
+            BH.oM.Geometry.Line extraLine2 = null;
+
+            List<BH.oM.Geometry.Point> endPoints = new List<BH.oM.Geometry.Point>();
+            List<BH.oM.Geometry.Point> midPoints = new List<BH.oM.Geometry.Point>();
+
+            // Get connectors
+            ConnectorManager connectorManager = mepCurve.ConnectorManager;
+            ConnectorSet connectorSet = connectorManager.Connectors;
+            List<Connector> c1Connectors = connectorSet.Cast<Connector>().ToList();
+            for (int i = 0; i < c1Connectors.Count; i++)
```

Comment:
This is a lot of `for` and `if` statements... I'm happy with it if it works (@michaelhoehn and others to confirm), but it might be good to refactor out to something slightly more maintainable.

Reply:
I 100% agree, this bugged my brain last week!  I'm adding more comments overall to make it easier to understand. Possible @pawelbaran has a suggestion on this since we discussed it a little bit?

Reply:
Following the chat with @rolyhudson, we believe that the location curves of MEP systems could be queried more efficiently with the use of graphs: parse the location curves and points of elements (linear objects and fittings respectively), then extract the relationships between them based on their `Connector` information, build the graph based on that and traverse it.

Consequently, I believe it is OK to merge this code as is (it seems to work great, simply hard to read and follow 😃) and replace it with a graph-based solution at some point. Happy to discuss further if any of you disagrees.

Reply:
I think that's a sensible approach @pawelbaran